src/data/move.ts

```diff
@@ -2875,6 +2875,164 @@ export class StatStageChangeAttr extends MoveEffectAttr {
   }
 }
 
+/**
+ * Attribute used to determine the Biome/Terrain-based secondary effect of Secret Power
+ */
+export class SecretPowerAttr extends MoveEffectAttr {
+  constructor() {
+    super(false);
+  }
+
+  /**
+   * Used to determine if the move should apply a secondary effect based on Secret Power's 30% chance
+   * @param user
+   * @param target
+   * @param move
+   * @param args
+   * @returns `true` if the move should | `false` if the roll fails and the move should not
+   */
+  override canApply(user: Pokemon, target: Pokemon, move: Move, args?: any[]): boolean {
+    this.effectChanceOverride = move.chance;
+    const moveChance = this.getMoveChance(user, target, move, this.selfTarget);
+    if (moveChance < 0 || moveChance === 100 || user.randSeedInt(100) < moveChance) {
+      this.effectChanceOverride = 100; // effectChanceOverride used in the application of the actual secondary effect
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  /**
+   * Used to apply the secondary effect to the target Pokemon
+   * @param user
+   * @param target
+   * @param move
+   * @param args
+   * @returns `true` if a secondary effect is successfully applied | 'false' if not
+   */
+  override apply(user: Pokemon, target: Pokemon, move: Move, args?: any[] | undefined): boolean | Promise<boolean> {
+    if (!super.apply(user, target, move, args)) {
+      return false;
+    }
+    let secondaryEffect: MoveEffectAttr;
+    const terrain = user.scene.arena.getTerrainType();
+    if (terrain !== TerrainType.NONE) {
+      secondaryEffect = this.determineTerrainEffect(terrain);
+    } else {
+      const biome = user.scene.arena.biomeType;
+      secondaryEffect = this.determineBiomeEffect(biome);
+    }
+    return secondaryEffect!.apply(user, target, move, []);
+  }
+
+  /**
+   * Determines the secondary effect based on terrain
+   * Takes precedence over biome
+   * Electric Terrain | Paralysis
+   * Misty Terrain    | SpAtk -1
+   * Grassy Terrain   | Sleep
+   * Psychic Terrain  | Speed -1
+   * @param terrain {@linkcode TerrainType}the current terrain
```

Comment:
```suggestion
   * Determines the secondary effect based on terrain.
   * Takes precedence over biome-based effects.
   * ```
   * Electric Terrain | Paralysis
   * Misty Terrain    | SpAtk -1
   * Grassy Terrain   | Sleep
   * Psychic Terrain  | Speed -1
   * ```
   * @param terrain - {@linkcode TerrainType} The current terrain
```

By adding the \`\`\`, it ensures the table stays properly formatted when viewing the tsdocs from hovering the function in VSCode. Tsdocs are converted to Markdown (like GitHub comments) when viewed, basically.